File: lessons/sci/modular/create_oboeru_from_wakaru.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
わかる編からおぼえる編を生成するスクリプト
- わかる編の text, choices, answer, qnum, tags, difficulty, asof をそのまま使用
- source は空文字列 "" にする
"""
import os
import json
import re
from pathlib import Path

# ディレクトリパス
WAKARU_DIR = Path(__file__).parent / 'wakaru'
OBOERU_DIR = Path(__file__).parent / 'oboeru'

# 除外するファイル
EXCLUDE_FILES = {
    'loader.js', 'script.js', 'index_modular.html', 'style.css',
    '*.html', '*.css', '*.md', '*.zip', '*sim.html', '*.backup*'
}

def get_wakaru_files():
    """わかる編の全ファイルを取得"""
    files = []
    for file in WAKARU_DIR.glob('*.js'):
        if file.name not in EXCLUDE_FILES and not any(
            file.name.endswith(ext) for ext in ['.backup', '.backup2', '.backup_final', '.backup_indent']
        ):
            files.append(file)
    return sorted(files)

def get_oboeru_filename_mapping():
    """既存のおぼえる編ファイル名のマッピングを作成"""
    mapping = {}
    
    # 既存のおぼえる編ファイルを確認
    for oboeru_file in OBOERU_DIR.glob('*.js'):
        if oboeru_file.name in ['loader.js', 'script.js']:
            continue
        
        name = oboeru_file.stem  # 拡張子なし
        
        # _oboeru サフィックスを削除してわかる編のファイル名を推測
        if name.endswith('_oboeru'):
            wakaru_name = name[:-7]  # '_oboeru' を削除
            mapping[wakaru_name] = name
        else:
            # サフィックスがない場合は同じ名前
            mapping[name] = name
    
    return mapping

def determine_oboeru_filename(wakaru_filename, mapping):
    """おぼえる編のファイル名を決定"""
    wakaru_name = wakaru_filename.stem
    
    if wakaru_name in mapping:
        return mapping[wakaru_name] + '.js'
    else:
        # マッピングがない場合は _oboeru サフィックスを追加
        return wakaru_name + '_oboeru.js'

def create_oboeru_from_wakaru(wakaru_file, oboeru_file):
    """わかる編からおぼえる編を作成"""
    try:
        # わかる編を読み込む
        with open(wakaru_file, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # window.questions を抽出
        # 簡易的な方法: eval を使わずに JSON を抽出
        match = re.search(r'window\.questions\s*=\s*(\[[\s\S]*\]);', content)
        if not match:
            print(f"[警告] {wakaru_file.name}: window.questions が見つかりません")
            return False
        
        questions_json = match.group(1)
        questions = json.loads(questions_json)
        
        # おぼえる編用に変換（source を空文字列に）
        oboeru_questions = []
        for q in questions:
            oboeru_q = {
                'qnum': q.get('qnum'),
                'text': q.get('text'),
                'choices': q.get('choices'),
                'answer': q.get('answer'),
                'source': '',  # 空文字列
                'tags': q.get('tags', []),
                'difficulty': q.get('difficulty', 1),
                'asof': q.get('asof', '2025-01-27')
            }
            oboeru_questions.append(oboeru_q)
        
        # おぼえる編ファイルを生成
        oboeru_content = 'window.questions = [\n'
        for i, q in enumerate(oboeru_questions):
            oboeru_content += '  {\n'
            oboeru_content += f'    "qnum": {json.dumps(q["qnum"], ensure_ascii=False)},\n'
            oboeru_content += f'    "text": {json.dumps(q["text"], ensure_ascii=False)},\n'
            oboeru_content += f'    "choices": {json.dumps(q["choices"], ensure_ascii=False)},\n'
            oboeru_content += f'    "answer": {q["answer"]},\n'
            oboeru_content += f'    "source": "",\n'
            oboeru_content += f'    "tags": {json.dumps(q["tags"], ensure_ascii=False)},\n'
            oboeru_content += f'    "difficulty": {q["difficulty"]},\n'
            oboeru_content += f'    "asof": {json.dumps(q["asof"], ensure_ascii=False)}\n'
            oboeru_content += '  }'
            if i < len(oboeru_questions) - 1:
                oboeru_content += ','
            oboeru_content += '\n'
        oboeru_content += '];\n'
        
        # ファイルを書き込む
        with open(oboeru_file, 'w', encoding='utf-8') as f:
            f.write(oboeru_content)
        
        return True
        
    except Exception as e:
        print(f"[エラー] ({wakaru_file.name}): {e}")
        import traceback
        traceback.print_exc()
        return False

def main():
    print("わかる編からおぼえる編を生成します...\n")
    
    # わかる編の全ファイルを取得
    wakaru_files = get_wakaru_files()
    print(f"わかる編ファイル数: {len(wakaru_files)}")
    
    # おぼえる編のファイル名マッピングを取得
    mapping = get_oboeru_filename_mapping()
    print(f"既存のおぼえる編マッピング: {len(mapping)}個\n")
    
    # 各ファイルを処理
    created = 0
    updated = 0
    errors = 0
    
    for wakaru_file in wakaru_files:
        oboeru_filename = determine_oboeru_filename(wakaru_file, mapping)
        oboeru_file = OBOERU_DIR / oboeru_filename
        
        existed = oboeru_file.exists()
        
        if create_oboeru_from_wakaru(wakaru_file, oboeru_file):
            if existed:
                print(f"[更新] {oboeru_filename} (元: {wakaru_file.name})")
                updated += 1
            else:
                print(f"[作成] {oboeru_filename} (元: {wakaru_file.name})")
                created += 1
        else:
            errors += 1
    
    print(f"\n完了!")
    print(f"  新規作成: {created}個")
    print(f"  更新: {updated}個")
    if errors > 0:
        print(f"  エラー: {errors}個")

if __name__ == '__main__':
    main()

File: lessons/sci/modular/create_oboeru_from_wakaru.js
#!/usr/bin/env node
/**
 * わかる編からおぼえる編を生成するスクリプト
 * - わかる編の text, choices, answer, qnum, tags, difficulty, asof をそのまま使用
 * - source は空文字列 "" にする
 */
const fs = require('fs');
const path = require('path');

// スクリプトのディレクトリを取得
const SCRIPT_DIR = __dirname || path.dirname(require.main.filename);
const WAKARU_DIR = path.join(SCRIPT_DIR, 'wakaru');
const OBOERU_DIR = path.join(SCRIPT_DIR, 'oboeru');

console.log('WAKARU_DIR:', WAKARU_DIR);
console.log('OBOERU_DIR:', OBOERU_DIR);

// 除外するファイル
const EXCLUDE_FILES = ['loader.js', 'script.js', 'index_modular.html', 'style.css'];

function getWakaruFiles() {
  const files = fs.readdirSync(WAKARU_DIR)
    .filter(file => file.endsWith('.js') && !EXCLUDE_FILES.includes(file))
    .filter(file => !file.includes('.backup'))
    .map(file => path.join(WAKARU_DIR, file));
  return files.sort();
}

function getOboeruFilenameMapping() {
  const mapping = {};
  
  // 既存のおぼえる編ファイルを確認
  if (fs.existsSync(OBOERU_DIR)) {
    const files = fs.readdirSync(OBOERU_DIR)
      .filter(file => file.endsWith('.js') && !EXCLUDE_FILES.includes(file));
    
    for (const file of files) {
      const name = path.basename(file, '.js');
      
      // _oboeru サフィックスを削除してわかる編のファイル名を推測
      if (name.endsWith('_oboeru')) {
        const wakaruName = name.slice(0, -7); // '_oboeru' を削除
        mapping[wakaruName] = name;
      } else {
        // サフィックスがない場合は同じ名前
        mapping[name] = name;
      }
    }
  }
  
  return mapping;
}

function determineOboeruFilename(wakaruFile, mapping) {
  const wakaruName = path.basename(wakaruFile, '.js');
  
  if (mapping[wakaruName]) {
    return mapping[wakaruName] + '.js';
  } else {
    // マッピングがない場合は _oboeru サフィックスを追加
    return wakaruName + '_oboeru.js';
  }
}

function createOboeruFromWakaru(wakaruFile, oboeruFile) {
  try {
    // わかる編を読み込む
    const content = fs.readFileSync(wakaruFile, 'utf-8');
    
    // window.questions を抽出して評価
    // 安全のため、vmモジュールを使用
    const vm = require('vm');
    const context = { window: {} };
    vm.createContext(context);
    vm.runInContext(content, context);
    
    const questions = context.window.questions;
    if (!Array.isArray(questions)) {
      console.log(`[警告] ${path.basename(wakaruFile)}: questions が配列ではありません`);
      return false;
    }
    
    // おぼえる編用に変換（source を空文字列に）
    const oboeruQuestions = questions.map(q => ({
      qnum: q.qnum,
      text: q.text,
      choices: q.choices,
      answer: q.answer,
      source: '', // 空文字列
      tags: q.tags || [],
      difficulty: q.difficulty || 1,
      asof: q.asof || '2025-01-27'
    }));
    
    // おぼえる編ファイルを生成
    let oboeruContent = 'window.questions = [\n';
    for (let i = 0; i < oboeruQuestions.length; i++) {
      const q = oboeruQuestions[i];
      oboeruContent += '  {\n';
      oboeruContent += `    "qnum": ${JSON.stringify(q.qnum)},\n`;
      oboeruContent += `    "text": ${JSON.stringify(q.text)},\n`;
      oboeruContent += `    "choices": ${JSON.stringify(q.choices)},\n`;
      oboeruContent += `    "answer": ${q.answer},\n`;
      oboeruContent += `    "source": "",\n`;
      oboeruContent += `    "tags": ${JSON.stringify(q.tags)},\n`;
      oboeruContent += `    "difficulty": ${q.difficulty},\n`;
      oboeruContent += `    "asof": ${JSON.stringify(q.asof)}\n`;
      oboeruContent += '  }';
      if (i < oboeruQuestions.length - 1) {
        oboeruContent += ',';
      }
      oboeruContent += '\n';
    }
    oboeruContent += '];\n';
    
    // ファイルを書き込む
    fs.writeFileSync(oboeruFile, oboeruContent, 'utf-8');
    
    return true;
    
  } catch (error) {
    console.log(`[エラー] (${path.basename(wakaruFile)}): ${error.message}`);
    console.error(error);
    return false;
  }
}

function main() {
  console.log('わかる編からおぼえる編を生成します...\n');
  
  // わかる編の全ファイルを取得
  const wakaruFiles = getWakaruFiles();
  console.log(`わかる編ファイル数: ${wakaruFiles.length}`);
  
  // おぼえる編のファイル名マッピングを取得
  const mapping = getOboeruFilenameMapping();
  console.log(`既存のおぼえる編マッピング: ${Object.keys(mapping).length}個\n`);
  
  // おぼえる編ディレクトリが存在しない場合は作成
  if (!fs.existsSync(OBOERU_DIR)) {
    fs.mkdirSync(OBOERU_DIR, { recursive: true });
  }
  
  // 各ファイルを処理
  let created = 0;
  let updated = 0;
  let errors = 0;
  
  for (const wakaruFile of wakaruFiles) {
    const oboeruFilename = determineOboeruFilename(wakaruFile, mapping);
    const oboeruFile = path.join(OBOERU_DIR, oboeruFilename);
    
    const existed = fs.existsSync(oboeruFile);
    
    if (createOboeruFromWakaru(wakaruFile, oboeruFile)) {
      if (existed) {
        console.log(`[更新] ${oboeruFilename} (元: ${path.basename(wakaruFile)})`);
        updated++;
      } else {
        console.log(`[作成] ${oboeruFilename} (元: ${path.basename(wakaruFile)})`);
        created++;
      }
    } else {
      errors++;
    }
  }
  
  console.log(`\n完了!`);
  console.log(`  新規作成: ${created}個`);
  console.log(`  更新: ${updated}個`);
  if (errors > 0) {
    console.log(`  エラー: ${errors}個`);
  }
}

main();

